### PR TITLE
Introduce enable_optimize_read_in_order in CHYT

### PR DESCRIPTION
**Pull request by achulkov2**
This PR introduces a new mode in CHYT that aims to mimic the behaviour of [optimize_read_in_order](https://clickhouse.com/docs/en/sql-reference/statements/select/order-by#optimization-of-data-reading) in native ClickHouse. This PR implements #451, albeit using a different approach.
It works by using a sorted pool to produce jobs of exponentially increasing/decreasing sizes, which turn into subqueries that are processed combined sequentially without additional sorting steps. This should allow queries described in #451 to execute without reading the whole table. 
Notable modifications with generic impact:
* Support for exponentially increasing job sizes in new sorted pool and related classes.
* Chunk slice fetcher is now used when creating a sorted pool in CHYT.
* New column in query log table exporter.

### CODE REVIEW

File: yt/chyt/server/config.cpp [hunk 1]
```diff
@@ -140,6 +140,10 @@ void TExecutionSettings::Register(TRegistrar registrar)
 
     registrar.Parameter("enable_min_max_filtering", &TThis::EnableMinMaxFiltering)
         .Default(true);
+
+    // TODO(achulkov2): Set to false by default. This is temporary to run all tests and verify that nothing fails.
+    registrar.Parameter("enable_optimize_read_in_order", &TThis::EnableOptimizeReadInOrder)
+        .Default(true);
```
Comment (achulkov2) [inline yt/chyt/server/config.cpp:146]: This is set to `true` temporarily, so that we run all CHYT tests in the import and see that nothing breaks.

File: yt/chyt/server/query_analyzer.cpp [hunk 2]
```diff
@@ -728,6 +744,125 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
...
+                THROW_ERROR_EXCEPTION_IF(
+                    orderByElement->nulls_direction_was_explicitly_specified,
+                    "Specified nulls direction is incompatible with enabled optimize_read_in_order mode; "
+                    "please turn it off via query settings, specify another nulls direction or use the default nulls direction");
+                YT_LOG_DEBUG(
+                    "Reversing ORDER BY nulls direction (ColumnName: %v, Direction: %v, NullsDirection: %v -> -1)",
+                    orderByElement->children.front()->getColumnName(),
+                    orderByElement->direction,
+                    orderByElement->nulls_direction);
+                orderByElement->nulls_direction = -1;
```
Comment (DimasKovas) [inline yt/chyt/server/query_analyzer.cpp:853]: Looks suspicious.
First, I'm not sure if we are allowed to change the query AST. I have no idea how CH code would handle this "unexpected change".
Second, it could change the query results compared to the execution without the optimization.

The second point would prevent this optimization from being turned on by default. If this behavior is desired, we could introduce a separate option to configure it.

File: yt/chyt/server/query_analyzer.cpp [hunk 3]
```diff
@@ -728,6 +744,125 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
...
+    // In YT, NULL values compare less than any other values. In ClickHouse, NULL value placement
+    // depends solely on the placement requested in the query, which defaults to NULLS LAST.
+    // Since we combine both YT and CH sorting in this mode, there is only one acceptable direction
+    // depending on the requested sort order. In cases when the unsupported option turns out to be
+    // the default (for ascending sorts), we tweak the query to avoid additional user turmoil.
+    if (commonDirection != EReadInOrderMode::None) {
+        for (const auto& orderByElementAst : selectQuery->orderBy()->children) {
+            auto orderByElement = orderByElementAst->as<DB::ASTOrderByElement>();
+
+            // The value of nulls_direction is equal to direction for NULLS LAST and to the opposite
```
Comment (DimasKovas) [inline yt/chyt/server/query_analyzer.cpp:839]: We can skip `required` columns. Don't need to required 'nulls first/last` for them

File: yt/chyt/server/query_analyzer.cpp [hunk 4]
```diff
@@ -728,6 +744,125 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
...
+
+        // Judging by CH code and experiments, ASTOrderByElement always has a single child.
+        if (std::ssize(orderByElement->children) != 1) {
+            THROW_ERROR_EXCEPTION("Query ORDER BY ast is malformed; "
+                "this is a bug; please contact cluster administrators");
+        }
+
+        // The name can contain functions. While it is possible to support tuples
+        // and even monotonic functions, I do not think there is much use.
+        auto columnName = orderByElement->children.front()->getColumnName();
```
Comment (DimasKovas) [inline yt/chyt/server/query_analyzer.cpp:797]: Need to check that the orderByElement is a `column` somehow before accessing it.

We don't want to break queries with an expression in the `order by` clause.

File: yt/chyt/server/query_analyzer.cpp [hunk 5]
```diff
@@ -728,6 +744,125 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
...
+                    orderByElement->nulls_direction);
+                orderByElement->nulls_direction = -1;
+                orderByElement->nulls_direction_was_explicitly_specified = true;
+            }
+        }
+    }
+    // Sadly, with NANs we are just screwed. In YT these values are larger than any others, but in
+    // CHYT they are always placed together with NULLs. These two approaches contradict each other.
+    // To warn about such cases without disabling this optimization for tables with floating-point
+    // key columns completely, we require to specify nulls direction explicitly as a green light.
```
Comment (DimasKovas) [inline yt/chyt/server/query_analyzer.cpp:861]: I offer to introduce a separate option for `float` behavior too, so we can safely turn `optimize_read_in_order` on by default

File: yt/chyt/server/subquery.cpp [hunk 6]
```diff
@@ -254,7 +255,7 @@ class TInputFetcher
                     EKeyConditionScale::TopLevelDataSlice,
                     dataSlice->LowerLimit().KeyBound,
                     dataSlice->UpperLimit().KeyBound,
-                    dataSlice->GetInputStreamIndex()).can_be_true;
+                    operandIndex).can_be_true;
```
Comment (DimasKovas) [inline yt/chyt/server/subquery.cpp:258]: I don't understand why this change is needed

File: yt/chyt/server/subquery.cpp [hunk 7]
```diff
@@ -426,7 +427,7 @@ class TInputFetcher
 
             for (auto& dataSlice : InputDataSlices_[tableIndex]) {
                 YT_VERIFY(!dataSlice->IsLegacy);
-                dataSlice->SetInputStreamIndex(operandIndex);
+                dataSlice->SetInputStreamIndex(tableIndex);
```
Comment (DimasKovas) [inline yt/chyt/server/subquery.cpp:430]: And I don't understand that one either

File: yt/yt/server/lib/chunk_pools/job_size_tracker.cpp [hunk 8]
```diff
@@ -143,6 +147,26 @@ class TJobSizeTracker
             return;
         }
 
+        if (Options_.LimitProgressionRatio && LimitProgressionLength_ < Options_.LimitProgressionLength) {
```
Comment (DimasKovas) [inline yt/yt/server/lib/chunk_pools/job_size_tracker.cpp:150]: I don't understand what is going on here, but if it works as expected, that's fine

**Review by DimasKovas — CHANGES_REQUESTED** [review]

File: yt/chyt/server/query_context.h [hunk 9]
```diff
@@ -278,3 +281,7 @@ void InvalidateCache(
 ////////////////////////////////////////////////////////////////////////////////
 
 } // namespace NYT::NClickHouseServer
+
+#define QUERY_CONTEXT_INL_H_
+#include "query_context-inl.h"
```
Comment (DimasKovas) [inline yt/chyt/server/query_context.h:286]: There is no `query_context-inl.h` in this PR

Comment (achulkov2) [reply yt/chyt/server/query_context.h:286]: That's a pity! Added id, so you can import again for tests.

File: yt/chyt/server/subquery.cpp (hunk 6 above)
Comment (achulkov2) [reply yt/chyt/server/subquery.cpp:258]: This is the consequence of the change in input stream index (it used to reflect the operand index and now it doesn't).

File: yt/chyt/server/query_context-inl.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+            key,
+            NYson::ConvertToYsonString(RuntimeVariables_->GetYson(key), NYson::EYsonFormat::Text),
+            NYson::ConvertToYsonString(value, NYson::EYsonFormat::Text));
+    }
+    RuntimeVariables_->Set(key, value);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+} // namespace NYT::NClickHouseServer
```
Comment (DimasKovas) [inline yt/chyt/server/query_context-inl.h:28]: '\n'

**Review by DimasKovas — APPROVED** [review]
To run tests

File: yt/chyt/server/subquery.cpp (hunk 7 above)
Comment (achulkov2) [reply yt/chyt/server/subquery.cpp:430]: In order for sorted pool to work properly, we need to attribute slices to correct input streams: the pool uses these streams to decide whether to use chunk slice fetcher (used for unversioned streams only) and in the end it validates that output slices from a single stream are actually sorted. I am pretty sure these indexes are also used while building jobs, otherwise it would be impossible to have multiple potentially intersecting sorted tables as input.

Previously, we used the operand index as input stream index, so all concatenated tables seemed like one big table to the pool. This has no chances of working, since this "one big table" is not actually sorted.

File: yt/chyt/server/query_analyzer.cpp (hunk 2 above)
Comment (achulkov2) [reply yt/chyt/server/query_analyzer.cpp:853]: For the first point, I don't know how to check whether it is okay to change it like that or not. I can say that first I only changed `nulls_direction` and it didn't have any effect, but once you also set `nulls_direction_was_explicitly_set` it starts having effect. There are tests for this and they pass. 
Also there are already some AST changes nearby in `Analyze`, so I think it should be fine...

Comment (achulkov2) [reply yt/chyt/server/query_analyzer.cpp:853]: As for the second point, my intent was exactly so that we could enable it by default and the users wouldn't have to manually write `NULLS FIRST` or `NULLS LAST` in their queries. 
I agree that this is all quite annoying. I am also concerned that optional key columns are probably not uncommon, since that is pretty much what you get by default.

I can introduce a separate option that would enable this behaviour and turn it off by default. If it is disabled, then we will throw an exception here instead of changing the AST. However, I still want to change the AST, because enabling this extra option is much easier for the user than thinking about the correct NULLS direction. It is also much easier to copy and paste an enabled option from query to query. 

What do you think?

File: yt/chyt/server/query_analyzer.cpp (hunk 4 above)
Comment (achulkov2) [reply yt/chyt/server/query_analyzer.cpp:797]: We shouldn't? A few lines below we compare this name to the name of the corresponding column (by index) in the schema. This should never be true for expressions, so we won't enable the optimization. There is one test for that.

File: yt/yt/server/lib/chunk_pools/job_size_tracker.cpp (hunk 8 above)
Comment (achulkov2) [reply yt/yt/server/lib/chunk_pools/job_size_tracker.cpp:150]: The class itself is quite complicated, yeah, it took me a while. You should double check the general intent and whether I computed parameters correctly. 
I spent quite some time debugging errors around this and idk if there are more mistakes.

File: yt/chyt/server/config.cpp (hunk 1 above)
Comment (achulkov2) [reply yt/chyt/server/config.cpp:146]: > Why not to use native ClickHouse's enable_optimize_read_in_order option?
Do we have the ability to enable/disable native CH-options for the whole clique?

Comment (achulkov2) [reply yt/chyt/server/config.cpp:146]: >Did you measure the worst possible performance degradation from this optimization? If the degradation is not significant, we can leave it on by default.

What is the best way to go about measuring this? Do we have perf tests? :)

File: yt/chyt/server/job_size_constraints.cpp [hunk 11]
```diff
@@ -49,11 +59,78 @@ IJobSizeConstraintsPtr CreateClickHouseJobSizeConstraints(
...
+        switch (readInOrderMode) {
+            case EReadInOrderMode::Forward:
+                // The first range is small!
+                dataWeightPerJob = startDataWeightPerJob;
+                break;
+            case EReadInOrderMode::Backward:
+                // We will be reading these ranges backwards!
+                jobSizeTrackerOptions.LimitProgressionRatio = 0.5;
+                // We need to perform an actual split for the optimization to work.
+                if (jobCount == 1) {
```
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/chyt/server/job_size_constraints.cpp:89]: No, because in forward mode we start with `startDataWeightPerJob`, which is small, instead of `dataWeightPerJob` computed above. If this small size is larger than the entire table, then that's fine and we only need one small job. 

The problem in backward mode is that we need to produce a few slices first before starting our exponential descent, and here we have no chance of doing that if we put everything in one single slice in the very beginning.

File: yt/chyt/server/config.cpp (hunk 1 above)
Comment (DimasKovas) [reply yt/chyt/server/config.cpp:146]: > Do we have the ability to enable/disable native CH-options for the whole clique?

Yes, we do. The default CH settings may be specified in the cliques's config like this: `{clickhouse={settings={enable_optimize_read_in_order=1}}}`

But afaik, it's turned on by default in CH

> What is the best way to go about measuring this? Do we have perf tests? :)

Unfortunately, we don't have perf tests yet. I think that a simple query on a large table, with all the rows being filtered out, would give us an idea of how bad the performance degradation might be.

**Review by DimasKovas — CHANGES_REQUESTED** [review]
There are quite a few tests failing after this change.

Here is the list:
```
test_atomicity.py::TestClickHouseAtomicity test_join_with_locks (gudqeit)
test_input_fetching.py::TestInputFetching[0] test_reading_dynamic_and_static_tables (gudqeit)
test_input_fetching.py::TestInputFetching[0] test_reading_multiple_dynamic_tables (gudqeit)
test_join_and_in.py::TestJoinAndInStress test_sorted_join_stress (max42)
test_join_and_in.py::TestJoinAndIn[0] test_complex_join_key (dakovalkov)
test_join_and_in.py::TestJoinAndIn[0] test_filter_joined_subquery (dakovalkov)
test_join_and_in.py::TestJoinAndIn[0] test_join_table_and_array_join (dakovalkov)
test_join_and_in.py::TestJoinAndIn[0] test_tricky_join (max42)
test_join_and_in.py::TestJoinAndIn[1] test_join_dynamic_tables_with_dynamic_stores (max42)
test_join_and_in.py::TestJoinAndIn[1] test_join_for_unsorted_tables (gudqeit)
test_join_and_in.py::TestJoinAndIn[1] test_sorted_join_simple (max42)
test_join_and_in.py::TestJoinAndIn[1] test_sorted_join_with_wider_key_condition (max42)
test_join_and_in.py::TestJoinAndIn[1] test_tricky_join2 (max42)
```

File: yt/chyt/server/query_analyzer.cpp (hunk 4 above)
Comment (DimasKovas) [reply yt/chyt/server/query_analyzer.cpp:797]: I expected it to throw an error if the `orderByElement` is not a column literal. But yeah, since the tests work as expected, it's not the case.

Anyway, imho, an explicit check that the element is a literal would show your intent clearer in the code.

**Comment by achulkov2** [comment]
> There are quite a few tests failing after this change.

I fixed a stupid bug, let's import again?

**Review by DimasKovas — APPROVED** [review]
To run tests

**Comment by achulkov2** [comment]
I fixed another bug in which I moved `queryAnalyzer.Prepare()` a bit too early in `getQueryProcessingStage` in `storage_distributor.cpp`.
I am also concerned that I introduced a call to `Analyze` there, which might actually change the query AST (in my new way, but also in already existing ways). 
Once we are finished with some of the discussions in the comments, I plan to separate the inference of `ReadInOrderMode`, which will happen in `ParseQuery` or `Prepare`, and the modification of ORDER BY in the AST, which will happen in `Analyze`. 
I also propose renaming `Analyze` into something more verbose of the fact that it can modify the AST.

**Review by DimasKovas — APPROVED** [review]
To run tests

File: yt/chyt/server/query_analyzer.cpp (hunk 2 above)
Comment (DimasKovas) [reply yt/chyt/server/query_analyzer.cpp:853]: I thought about it one more time and now I'm pretty sure that this type of change is illegal.

For example, let's look at this query:

```sql
select * from (
    select * from "//sorted_table"
    order by key
    limit 5
    union all
    select * from "//unsorted_table"
    order by key
    limit 5
) order by key
```

In this case ClickHouse may substitute the last `order by key` phase with `sorted merge`, because both subqueries have already been sorted by the key.

But if you change the order of nulls, two subqueries would be sorted differently, and sorted merge may either throw an error or return an incorrect result.

The same applies to NaNs - I'm not sure that it's a good idea to have this logic at all, even under an option.

Comment (achulkov2) [reply yt/chyt/server/query_analyzer.cpp:853]: The problem is that without this logic we might get incorrect results when read_in_order is enabled. What should we do?

File: yt/chyt/server/query_analyzer.cpp [hunk 12]
```diff
@@ -728,6 +744,125 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
...
+        for (const auto& orderByElementAst : selectQuery->orderBy()->children) {
+            auto orderByElement = orderByElementAst->as<DB::ASTOrderByElement>();
+
+            // The value of nulls_direction is equal to direction for NULLS LAST and to the opposite
+            // of direction for NULLS FIRST.
+            // For ASC (d = 1), we need NULLS FIRST (nd = -d = -1).
+            // For DSC (d = -1), we need NULLS LAST (nd =  d = -1).
+            if (orderByElement->nulls_direction != -1) {
+                THROW_ERROR_EXCEPTION_IF(
+                    orderByElement->nulls_direction_was_explicitly_specified,
```
Comment (DimasKovas) [inline yt/chyt/server/query_analyzer.cpp:845]: We have discussed that there should be no errors here - just disable the optimization if the order by clause is incompatible.

File: yt/chyt/server/query_analyzer.cpp (hunk 2 above)
Comment (achulkov2) [reply yt/chyt/server/query_analyzer.cpp:853]: New approach: we do not change AST in any way, but add two additional options `assume_no_null_keys` and `assume_no_nan_keys`.
We check nulls direction for a key column if it could have either nulls or nans, but not both at the same time (only for floats), and only optimize if it is correct. If a float key column could have both nulls and nans we never optimize.

File: yt/chyt/server/query_analyzer.cpp (hunk 12 above)
Comment (achulkov2) [reply yt/chyt/server/query_analyzer.cpp:845]: Removed this, see other comments.

File: yt/chyt/server/config.cpp (hunk 1 above)
Comment (achulkov2) [reply yt/chyt/server/config.cpp:146]: Since we want to disable the option by default for now, I will keep our separate option. I don't think it is easy (or that we should) disable a native CH option from our code. 

If we decide to enable it by default, we should probably deprecate it in favor of the native option (which is turned on by default).

Comment (DimasKovas) [reply yt/chyt/server/config.cpp:146]: Ok

File: yt/chyt/server/query_analyzer.cpp (hunk 4 above)
Comment (achulkov2) [reply yt/chyt/server/query_analyzer.cpp]: Changed it to a cast to `ASTIdentifier`. It is even better, because it works properly with aliases introduced by `WITH`!

**Comment by achulkov2** [comment]
I have implemented what we have talked about, if there are no more issues, let's import the PR, wait for tests to pass internally, then change the default to `false` and import and merge.

**Review by DimasKovas — APPROVED** [review]
Tu run tests

File: yt/chyt/server/query_context-inl.h [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+////////////////////////////////////////////////////////////////////////////////
+
+template <class T>
+void TQueryContext::SetRuntimeVariable(const TString& key, const T& value)
+{
+    auto guard = Guard(QueryLogLock_);
+    YT_VERIFY(RuntimeVariables_);
+    if (RuntimeVariables_->Contains(key)) {
+        YT_LOG_WARNING(
```
Comment (DimasKovas) [inline yt/chyt/server/query_context-inl.h:17]: Just in case:
TQueryContext is shared between all subqueries in the query.
E.g. if you have query like this:
```sql
SELECT * FROM t1
ORDER BY key
LIMIT 10
UNION ALL
SELECT * FROM t2
ORDER BY key
LIMIT 10
```
You will override the runtime variables when executing the second subquery and trigger this `YT_LOG_WARNING`

**Review by DimasKovas — APPROVED** [review]
LGTM

Please, run `ya test --style` in `yt/chyt/server/tests` - there are a log of `flake8` issues in your tests.

Unfortunately, CHYT tests weren't run internally because of an infrastructure issue. I've just restarted the check. If it passes, let's change default to `false` and merge it

File: yt/chyt/server/query_context-inl.h (hunk 13 above)
Comment (achulkov2) [reply yt/chyt/server/query_context-inl.h:17]: After our discussion, I don't think it is worth investing into fixing this. Changed to YT_LOG_DEBUG, so that we could see it in the log.

File: yt/chyt/server/query_analyzer.cpp [hunk 14]
```diff
@@ -728,6 +754,115 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
...
+    }
+
+    auto commonDirection = EReadInOrderMode::None;
+
+    // Columns from the ORDER BY clause must form a prefix of the table's primary key.
+    if (std::ssize(selectQuery->orderBy()->children) > schema->GetKeyColumnCount()) {
+        return;
+    }
+
+    std::vector<TString> floatKeyColumnsWithoutExplicitNullsDirection;
```
[the remark this answers is not in the document]
Comment (achulkov2) [reply yt/chyt/server/query_analyzer.cpp:794]: Forgot to fix this in the last commit, done.

**Comment by DimasKovas** [comment]
I don't see any relevant test failures. Let's change the default to `false` and merge

**Comment by achulkov2** [comment]
> I don't see any relevant test failures. Let's change the default to `false` and merge

Done.

**Review by DimasKovas — APPROVED** [review]
